Userland/Libraries/LibDisassembly/riscv64/Encoding.h

```diff
@@ -0,0 +1,243 @@
+};
+struct RawBType {
+    i32 imm;
+    Register rs2;
+    Register rs1;
+    u8 funct3;
+    u8 opcode;
+
+    static RawBType parse(u32 instruction);
+};
+struct RawUType {
+    i32 imm;
+    Register rd;
+    u8 opcode;
+
+    static RawUType parse(u32 instruction);
+};
+struct RawJType {
+    i32 imm;
+    Register rd;
+    u8 opcode;
+
+    static RawJType parse(u32 instruction);
+};
+struct RawR4Type {
+    Register rs3;
+    u8 fmt;
+    Register rs2;
+    Register rs1;
+    u8 funct3;
+    Register rd;
+    u8 opcode;
+
+    static RawR4Type parse(u32 instruction);
+};
+struct RawCRType {
+    u8 funct4;
+    Register rd_or_rs1;
+    Register rs2;
+    u8 opcode;
+
+    static RawCRType parse(u16 instruction);
+};
+struct RawCIType {
+    u8 funct3;
+    i32 imm;
+    Register rd_or_rs1;
+    u8 opcode;
+
+    static RawCIType parse(u16 instruction);
+};
+struct RawCSSType {
+    u8 funct3;
+    i32 imm;
+    Register rs2;
+    u8 opcode;
+
+    static RawCSSType parse(u16 instruction);
+};
+struct RawCIWType {
+    u8 funct3;
+    i32 imm;
+    Register rd;
+    u8 opcode;
+
+    static RawCIWType parse(u16 instruction);
+};
+struct RawCLType {
+    u8 funct3;
+    i32 imm;
+    Register rs1;
+    Register rd;
+    u8 opcode;
+
+    static RawCLType parse(u16 instruction);
+};
+struct RawCSType {
+    u8 funct3;
+    i32 imm;
+    Register rs1;
+    Register rs2;
+    u8 opcode;
+
+    static RawCSType parse(u16 instruction);
+};
+struct RawCAType {
+    u8 funct6;
+    Register rs1_or_rd;
+    u8 funct2;
+    Register rs2;
+    u8 opcode;
+
+    static RawCAType parse(u16 instruction);
+};
+struct RawCBType {
+    u8 funct3;
+    i32 offset;
+    u8 funct2;
+    Register rs1;
+    u8 opcode;
+
+    static RawCBType parse(u16 instruction);
+};
+struct RawCJType {
+    u8 funct3;
+    i32 jump_target;
+    u8 opcode;
+
+    static RawCJType parse(u16 instruction);
+};
+
+// Table 24.1
+enum class MajorOpcode : u8 {
+    LOAD = 0b00000'11,
+    STORE = 0b01000'11,
+    MADD = 0b10000'11,
+    BRANCH = 0b11000'11,
+    LOAD_FP = 0b00001'11,
+    STORE_FP = 0b01001'11,
+    MSUB = 0b10001'11,
+    JALR = 0b11001'11,
+    custom_0 = 0b00010'11,
+    custom_1 = 0b01010'11,
+    NMSUB = 0b10010'11,
+    reserved_0 = 0b11010'11,
+    MISC_MEM = 0b00011'11,
+    AMO = 0b01011'11,
+    NMADD = 0b10011'11,
+    JAL = 0b11011'11,
+    OP_IMM = 0b00100'11,
+    OP = 0b01100'11,
+    OP_FP = 0b10100'11,
+    SYSTEM = 0b11100'11,
+    AUIPC = 0b00101'11,
+    LUI = 0b01101'11,
+    reserved_1 = 0b10101'11,
+    reserved_2 = 0b11101'11,
+    OP_IMM_32 = 0b00110'11,
+    OP_32 = 0b01110'11,
+    custom_2_rv128 = 0b10110'11,
+    custom_3_rv128 = 0b11110'11,
+};
+
+// Table 16.4
+// Stored in the funct3 field and lowest 2 bits of compressed instructions.
+// Note that we always decode the RV64C version, but the names are as per specification and refer to all three variants.
+enum class CompressedOpcode : u8 {
+    ADDI4SPN = 0b000'00,
```

Comment:
same

Reply:
as you may have guesses, lowest bits 00 means compressed instruction (size 16 bits) :)